analysis/src/SignatureHelp.ml

```diff
@@ -0,0 +1,216 @@
+                 unlabelledArgCount := currentUnlabelledArgCount + 1;
+                 (* An argument without a label is just the expression, so we can use that. *)
+                 if arg.exp.pexp_loc |> Loc.hasPos ~pos:posBeforeCursor then
+                   Some (Unlabelled currentUnlabelledArgCount)
+                 else None
+               | Some {name; posStart; posEnd} -> (
+                 (* Check for the label identifier itself having the cursor *)
+                 match
+                   pos |> CursorPosition.classifyPositions ~posStart ~posEnd
+                 with
+                 | HasCursor -> Some (Labelled name)
+                 | NoCursor | EmptyLoc -> (
+                   (* If we're not in the label, check the exp. Either the exp
+                      exists and has the cursor. Or the exp is a parser recovery
+                      node, in which case we assume that the parser recovery
+                      indicates that the cursor was here. *)
+                   match
+                     ( arg.exp.pexp_desc,
+                       arg.exp.pexp_loc
+                       |> CursorPosition.classifyLoc ~pos:posBeforeCursor )
+                   with
+                   | Pexp_extension ({txt = "rescript.exprhole"}, _), _
+                   | _, HasCursor ->
+                     Some (Labelled name)
+                   | _ -> None)))
+      in
+      setFound (argAtCursor, exp, extractedArgs)
+    | _ -> ());
+    Ast_iterator.default_iterator.expr iterator expr
+  in
+  let iterator = {Ast_iterator.default_iterator with expr} in
+  let parser = Res_driver.parsingEngine.parseImplementation ~forPrinter:false in
+  let {Res_driver.parsetree = structure} = parser ~filename:currentFile in
+  iterator.structure iterator structure |> ignore;
+  match !foundFunctionApplicationExpr with
+  | Some (argAtCursor, exp, _extractedArgs) -> (
+    (* Not looking for the cursor position after this, but rather the target function expression's loc. *)
+    let pos = exp.pexp_loc |> Loc.end_ in
+    match findFunctionType ~currentFile ~debug ~path ~pos with
+    | Some (args, name, docstring, type_expr) ->
+      if debug then
+        Printf.printf "argAtCursor: %s\n"
+          (match argAtCursor with
+          | None -> "none"
+          | Some (Labelled name) -> "~" ^ name
+          | Some (Unlabelled index) -> "unlabelled<" ^ string_of_int index ^ ">");
+
+      (* The LS protocol wants us to send both the full type signature (label) that the end user sees as the signature help, and all parameters in that label
+         in the form of a list of start/end character offsets. We'll leverage the parser to figure the offsets out by parsing the label, and extract the
+         offsets from the parser. *)
+
+      (* Put together a label here that both makes sense to show to the end user in the signature help, but also can be passed to the parser. *)
+      let label = "let " ^ name ^ ": " ^ Shared.typeToString type_expr in
```

Comment:
Can you walk through the reason for printing a type expr, then parse it back right away?
Presumably it reuse some existing logic, but is it really worth it (potential for mismatch between printing and parsing)?

Reply:
The editor wants 2 things to be able to show the signature help;
1) The _string_ to show. This should be the full type definition string for the function we're looking at.
2) Offsets for all available arguments inside of that exact string, and what argument (if any) is currently active.

Producing 1) is easy obviously. 2) is a bit more tricky, since I need the exact offsets for that particular type string. So, I'm using the parser to tell me the precise location of each argument (so I can calculate offsets). And I ensure those locs/offsets are correct by using the parser to parse the exact string with the type definition that the editor will show, so the parser can tell me the precise location of each argument inside of that.

Are you following what I mean?

Reply:
Got it. Doing that in a more principled way would require instrumenting the printer, which sounds like more work and a one-off application.
Would you add a little line of comment there?

Reply:
This comment is a line above:
```
(* The LS protocol wants us to send both the full type signature (label) that the end user sees as the signature help, and all parameters in that label
         in the form of a list of start/end character offsets. We'll leverage the parser to figure the offsets out by parsing the label, and extract the
         offsets from the parser. *)
```

Is that enough or should I extend that more?

Reply:
Done.